subprojects/core/src/main/java/org/gradle/api/internal/project/taskfactory/DefaultTaskClassInfoStore.java

```diff
@@ -26,110 +26,79 @@
-                        continue;
-                    }
-                }
-                if (previousDeclaringClass == declaringClass) {
-                    throw new GradleException(String.format(
-                        "Cannot use @TaskAction annotation on multiple overloads of method %s.%s()",
-                        declaringClass.getSimpleName(), method.getName()
-                    ));
-                } else if (previousDeclaringClass != null) {
-                    continue;
-                }
-                if (taskActionFactory instanceof IncrementalTaskActionFactory) {
-                    if (foundIncrementalTaskActionFactory != null) {
-                        throw new GradleException(String.format("Cannot have multiple @TaskAction methods accepting an %s parameter.", InputChanges.class.getSimpleName()));
-                    }
-                    foundIncrementalTaskActionFactory = (IncrementalTaskActionFactory) taskActionFactory;
-                    continue;
-                }
-                taskActionFactoriesBuilder.add(taskActionFactory);
+        typeMetadata.getMethodsMetadata().stream().filter(methodMetadata -> methodMetadata.getAnnotation(TaskAction.class).isPresent()).forEach(methodMetadata -> {
+            MethodMetadata alreadySeen = methods.put(methodMetadata.getMethodName(), methodMetadata);
+            if (alreadySeen != null) {
+                throw new GradleException(String.format(
+                    "Cannot use @TaskAction annotation on multiple overloads of method %s.%s()",
+                    typeMetadata.getType().getSimpleName(), methodMetadata.getMethod().getName()
+                ));
             }
-        }
-        if (foundIncrementalTaskActionFactory != null) {
-            taskActionFactoriesBuilder.add(foundIncrementalTaskActionFactory);
-        }
 
-        return new TaskClassInfo(taskActionFactoriesBuilder.build(), cacheable, reasonNotToTrackState);
-    }
-
-    @Nullable
-    private static TaskActionFactory createTaskAction(Class<? extends Task> taskType, final Method method) {
-        if (method.getAnnotation(TaskAction.class) == null) {
-            return null;
-        }
-        Class<?> declaringClass = method.getDeclaringClass();
-        if (Modifier.isStatic(method.getModifiers())) {
-            throw new GradleException(String.format("Cannot use @TaskAction annotation on static method %s.%s().",
-                declaringClass.getSimpleName(), method.getName()));
-        }
-        final Class<?>[] parameterTypes = method.getParameterTypes();
-        if (parameterTypes.length > 1) {
-            throw new GradleException(String.format(
-                "Cannot use @TaskAction annotation on method %s.%s() as this method takes multiple parameters.",
-                declaringClass.getSimpleName(), method.getName()));
-        }
+            // TODO These validations should be done as validation in TaskActionAnnotationHandler and surfaced as problems
```

Comment:
💭 TODO - Did you mean to do this prior to merging?  These validations are specific to this one type of annotation.

Reply:
The problem here is that doing them as validations moves the point that the exceptions are thrown to whenever the type is queried for problems.  This means that either 1) we move the exception to when task validation occurs (during execution) rather than during task creation or 2) we query for problems during task creation which moves task _property_ problems to be thrown during task creation instead of during task validation (during execution).  I wasn't sure which of these was better (I _think_ it would be ok to just move these to task validation time) so I figured we could just defer this decision until later (i.e. there's no reason we have to change this now).